File: app/src/main/java/com/wuruoye/know/util/sql/table/RecordTypeTable.java
package com.wuruoye.know.util.sql.table;

import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import java.util.ArrayList;
import java.util.List;

public class RecordTypeTable implements Table {
    public static final String NAME = "record_type";
    public static final String TITLE = "title";
    public static final String ITEMS = "items";
    public static final String CREATE_TIME = "create_time";
    public static final String UPDATE_TIME = "update_time";

    private int id;
    private String title;
    private String items;
    private long createTime;
    private long updateTime;

    public RecordTypeTable(int id, String title, String items, long createTime, long updateTime) {
        this.id = id;
        this.title = title;
        this.items = items;
        this.createTime = createTime;
        this.updateTime = updateTime;
    }

    public static void create(SQLiteDatabase db) {
        db.execSQL("create table " + NAME + " (" +
                "id integer primary key autoincrement, " +
                TITLE + " text, " +
                ITEMS + " text, " +
                CREATE_TIME + " integer, " +
                UPDATE_TIME + " integer " +
                ")");
    }

    public static List<RecordTypeTable> queryAll(SQLiteDatabase db) {
        List<RecordTypeTable> result = new ArrayList<>();
        Cursor cursor = db.query(NAME, null, null,
                null, null, null, CREATE_TIME);

        try {
            cursor.moveToFirst();
            while (!cursor.isAfterLast()) {
                int id = cursor.getInt(0);
                String title = cursor.getString(1);
                String items = cursor.getString(2);
                long createTime = cursor.getLong(3);
                long updateTime = cursor.getLong(4);
                result.add(new RecordTypeTable(id, title, items, createTime, updateTime));
            }
        } finally {
            cursor.close();
        }

        return result;
    }

    @Override
    public boolean save(SQLiteDatabase db) {
        return false;
    }

    @Override
    public boolean delete(SQLiteDatabase db) {
        return false;
    }

    @Override
    public boolean update(SQLiteDatabase db) {
        return false;
    }

    public int getId() {
        return id;
    }

    public String getTitle() {
        return title;
    }

    public String getItems() {
        return items;
    }

    public long getCreateTime() {
        return createTime;
    }

    public long getUpdateTime() {
        return updateTime;
    }
}

File: app/src/main/java/com/wuruoye/know/util/sql/table/RecordTypeTable.kt
package com.wuruoye.know.util.sql.table

import android.content.ContentValues
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import com.wuruoye.know.model.beans.RecordType
import java.util.*

class RecordTypeTable(id: Int,
                      val title: String,
                      val items: String,
                      val createTime: Long,
                      val updateTime: Long) : Table(id) {

    constructor(recordType: RecordType, items: String): this(recordType.id, recordType.title,
            items, recordType.createTime, recordType.updateTime)

    override fun save(db: SQLiteDatabase): Int {
        return if (id < 0) {
            db.insert(NAME, null, contentValues()).toInt()
        } else {
            db.update(NAME, contentValues(), "id=?", arrayOf(id.toString()))
            id
        }
    }

    override fun delete(db: SQLiteDatabase): Int {
        return db.delete(NAME, "id=?", arrayOf(id.toString()))
    }

    override fun contentValues(): ContentValues {
        val values = ContentValues()
        values.put(TITLE, title)
        values.put(ITEMS, items)
        values.put(CREATE_TIME, createTime)
        values.put(UPDATE_TIME, updateTime)
        return values
    }

    companion object {
        val NAME = "record_type"
        val TITLE = "title"
        val ITEMS = "items"
        val CREATE_TIME = "create_time"
        val UPDATE_TIME = "update_time"
        val DESC = " desc"

        fun create(db: SQLiteDatabase) {
            db.execSQL("create table " + NAME + " (" +
                    "id integer primary key autoincrement, " +
                    TITLE + " text, " +
                    ITEMS + " text, " +
                    CREATE_TIME + " integer, " +
                    UPDATE_TIME + " integer " +
                    ")")
        }

        fun queryAll(db: SQLiteDatabase): List<RecordTypeTable> {
            val result = ArrayList<RecordTypeTable>()
            val cursor = db.query(NAME, null, null, null,
                    null, null, CREATE_TIME + DESC)

            cursor.use { c ->
                c.moveToFirst()
                while (!c.isAfterLast) {
                    result.add(fromCursor(c))
                    c.moveToNext()
                }
            }

            return result
        }

        fun query(db: SQLiteDatabase, id: Int): RecordTypeTable {
            val cursor = db.query(NAME, null, "id=?", arrayOf(id.toString()),
                    null, null, CREATE_TIME + DESC)
            cursor.use { c ->
                c.moveToFirst()
                return fromCursor(c)
            }
        }

        private fun fromCursor(cursor: Cursor): RecordTypeTable {
            val id = cursor.getInt(0)
            val title = cursor.getString(1)
            val items = cursor.getString(2)
            val createTime = cursor.getLong(3)
            val updateTime = cursor.getLong(4)
            return RecordTypeTable(id, title, items, createTime, updateTime)
        }
    }
}
